### DISCUSSION TITLE
Possible source generator fix

### DISCUSSION THREAD

**Discussion by Briaoeuidhtns** [christianhelle/refitter#495] [Ideas]
I came across [this recent comment](https://github.com/dotnet/roslyn/issues/57239#issuecomment-2266239308) that suggests a workaround for dependent source generators. I can look into it myself, but I want to make sure it's not something that's been tried already.

**Comment by christianhelle** [comment]
@Briaoeuidhtns No, I haven't tried to re-run the generators that should be aware of the generated code, from your this generator

It's worth a try!

**Comment by christianhelle** [comment]
@Briaoeuidhtns Did it work?

**Comment by christianhelle** [comment]
@Briaoeuidhtns I got a bit curious, and I also gave it a shot. As far as I can see, this would only work if Refit exposed [InterfaceStubGenerator](https://github.com/reactiveui/refit/blob/ef9b8305c35eb0644e31c33e43a7d0acaec81eec/InterfaceStubGenerator.Shared/InterfaceStubGenerator.cs) in the Refit package. This is, unfortunately, not the case right now. We could try to duplicate it, but that would be a bit counter-productive as we will need to keep it in sync with Refits implementation

**Reply by Briaoeuidhtns** [reply]
Thanks, I hadn't had a chance to check yet. Is it worth making an issue for Refit to see if they'd expose it?

**Reply by christianhelle** [reply]
We should investigate more before we involve the Refit team. I would like to see a working proof-of-concept that was suggested in https://github.com/dotnet/roslyn/issues/57239#issuecomment-2266239308

When I gave it a shot, I quickly ran into a wall regarding compilation units being immutable and the code that Refitter generated wasn't picked up by the [InterfaceStubGenerator](https://github.com/reactiveui/refit/blob/ef9b8305c35eb0644e31c33e43a7d0acaec81eec/InterfaceStubGenerator.Shared/InterfaceStubGenerator.cs) correctly. 

If I (or someone else, hint hint...) get to a working proof-of-concept after some minor tweaks to [InterfaceStubGenerator](https://github.com/reactiveui/refit/blob/ef9b8305c35eb0644e31c33e43a7d0acaec81eec/InterfaceStubGenerator.Shared/InterfaceStubGenerator.cs) without breaking existing Refit functionality then we could do a pull request on Refit

**Comment by christianhelle** [comment]
@Briaoeuidhtns I have started building another solution that involves MSBuild custom tasks that would generate the code into files on disk pre-compile time. It has a bit of the same solution as the source generator where it recursively scans for `.refitter` files in the folder containing the `.csproj` file and runs them with Refitter

The build output of the MSBuild task looks something like this:

```
   1:7>Target "BeforeBuild" in file "/Users/christianhelle/.nuget/packages/refitter.msbuild/1.0.0/build/Refitter.MSBuild.targets" from project "/Users/christianhelle/projects/christianhelle/refitter/test/MSBuild/Refitter.MSBuild.Tests.csproj" (target "Build" depends on it):
       Assembly loaded during TaskRun: Refitter.MSBuild, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null (location: /Users/christianhelle/.nuget/packages/refitter.msbuild/1.0.0/tasks/netstandard2.0/Refitter.MSBuild.dll, MVID: 833b493e-0849-4cbf-8444-60b6ac753f24, AssemblyLoadContext: MSBuild plugin /Users/christianhelle/.nuget/packages/refitter.msbuild/1.0.0/tasks/netstandard2.0/Refitter.MSBuild.dll)
       Using "RefitterGenerateTask" task from assembly "/Users/christianhelle/.nuget/packages/refitter.msbuild/1.0.0/build/../tasks/netstandard2.0/Refitter.MSBuild.dll".
       Task "RefitterGenerateTask"
         Starting RefitterGenerateTask
         Looking for .refitter files under /Users/christianhelle/projects/christianhelle/refitter/test/MSBuild
         Found 1 .refitter files...
         Processing /Users/christianhelle/projects/christianhelle/refitter/test/MSBuild/petstore.refitter
```

The difference in this approach is that code generation requires a new build, whereas source generators can incrementally generate code in the background as you make changes in Visual Studio. I don't always use Visual Studio, so I have a habit of building the solution from the CLI which makes this solution suitable for me
